File: pizzeria/pizzeria-core/src/main/java/io/github/joxit/pizzeria/mapper/PizzaMapper.java
package io.github.joxit.pizzeria.mapper;

import io.github.joxit.pizzeria.dto.PizzaDTO;
import io.github.joxit.pizzeria.model.Ingredient;
import io.github.joxit.pizzeria.model.Pizza;
import io.github.joxit.pizzeria.model.PizzaCustom;
import io.github.joxit.pizzeria.model.PizzaEager;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Jones Magloire @Joxit
 * @since 2017-11-01
 */
@Component
public class PizzaMapper {
  public PizzaDTO modelToDTO(Pizza pizza) {
    PizzaDTO pizzaDTO = new PizzaDTO();
    pizzaDTO.setName(pizza.getName());
    pizzaDTO.setPrix(pizza.getPrix());
    List<String> list = new ArrayList<>();
    for (Ingredient ingredient : pizza.getIngredients()) {
      String name = ingredient.getName();
      list.add(name);
    }
    pizzaDTO.setIngredients(list);
    return pizzaDTO;
  }

  public PizzaDTO modelToDTO(PizzaEager pizza) {
    PizzaDTO pizzaDTO = new PizzaDTO();
    pizzaDTO.setName(pizza.getName());
    pizzaDTO.setPrix(pizza.getPrix());
    pizzaDTO.setIngredients(pizza.getIngredients().stream().map(Ingredient::getName).collect(Collectors.toList()));
    return pizzaDTO;
  }

  public PizzaDTO modelToDTO(PizzaCustom pizza) {
    PizzaDTO pizzaDTO = new PizzaDTO();
    pizzaDTO.setName(pizza.getName());
    pizzaDTO.setPrix(pizza.getPrix());
    pizzaDTO.setIngredients(pizza.getIngredients().stream().map(Ingredient::getName).collect(Collectors.toList()));
    return pizzaDTO;
  }
}

File: pizzeria/pizzeria-core/src/main/kotlin/io/github/joxit/pizzeria/mapper/PizzaMapper.kt
package io.github.joxit.pizzeria.mapper

import io.github.joxit.pizzeria.dto.PizzaDTO
import io.github.joxit.pizzeria.model.Ingredient
import io.github.joxit.pizzeria.model.Pizza
import io.github.joxit.pizzeria.model.PizzaCustom
import io.github.joxit.pizzeria.model.PizzaEager
import org.springframework.stereotype.Component

/**
 * Old Java version : https://github.com/Joxit/IG-Master2/tree/promo-2021/pizzeria/pizzeria-core/src/main/java/io/github/joxit/pizzeria
 * @author Jones Magloire @Joxit
 * @since Java: 2017-11-01
 * @since Kotlin: 2022
 */
@Component
class PizzaMapper {
  fun modelToDTO(pizza: Pizza): PizzaDTO {
    val ingredients = mutableListOf<String>()
    // Boucle traditionnelle
    for (ingredient in pizza.ingredients) {
      ingredients.add(ingredient.name)
    }
    return PizzaDTO(
      pizza.name,
      pizza.price,
      ingredients
    )
  }

  fun modelToDTO(pizza: PizzaEager): PizzaDTO {
    return PizzaDTO(
      pizza.name,
      pizza.price,
      // Fonction lambda map avec it qui représente l'objet actuel
      pizza.ingredients.map { it.name }
    )
  }

  fun modelToDTO(pizza: PizzaCustom): PizzaDTO {
    return PizzaDTO(
      pizza.name,
      pizza.price,
      // Fonction lambda map avec référence sur méthode/champs
      pizza.ingredients.map(Ingredient::getName)
    )
  }
}